### ISSUE TITLE
Terminal focus key binding is not working correctly after update to 1.6

### ISSUE THREAD

**Issue by apbendi** [Microsoft/vscode#13511]
- VSCode Version: 1.6
- OS Version: OS X 10.11.6

Steps to Reproduce:
1. Add the following to your `keybindings.json`
2. Observe that it is broken

```
[{
    "key": "cmd+`",
    "command": "workbench.action.terminal.focus",
    "when": "editorTextFocus"
}]
```

Seems like a possible regression of #11714. Starting to get spooked by that update banner 😱

**Comment by isidorn** [comment]
Adding important since it might be a regression

**Comment by apbendi** [comment]
thanks @isidorn

**Comment by Tyriar** [comment]
It's working just fine for me. Note that this is overriding the key combination when `editorTextFocus` only, so pressing cmd+` the first time (when the editor is focused) it will focus the terminal, and when pressing cmd+` the second time (when the terminal is focused) it will toggle the terminal as that's the default.

**Comment by apbendi** [comment]
hey @Tyriar, with the above keybinding I used to be able to press cmd-backtick any time I was in an editor and jump to the terminal. If the terminal was not open, it would open it. If it was, it would jump there. Afterwards, I could press cmd-1 to return to the first editor window, or cmd-j to close the terminal.

This seems to no longer work. Pressing cmd-backtick doesn't do anything when in the editor. However, I just closed down VSCode and restarted, and upon trying again it worked the first time I pressed it, but failed on all subsequent attempts to use. When I restarted the editor again, I noticed the same behavior: it worked the first time but never again, regardless of whether I closed the terminal, etc... 

Are you sure this is working for you, or is it possibly somehow intermittent?

**Comment by apbendi** [comment]
Hey @petecorreia I noticed you 👍 'd my original issue- are you also still seeing this as broken?

**Comment by Tyriar** [comment]
I just tried again and it seems to work just fine. Note that I was using ctrl+` on a Linux machine. I just switched over to Mac, the toggle terminal keybinding has always been ctrl+` on mac, not cmd+`. If you want cmd+` to close the terminal when you're done you need to bind that as well?

**Comment by apbendi** [comment]
@Tyriar My keybinding is specifically for **cmd**+backtick on a Mac. ctrl+backtick works fine but I wanted a command that lets me jump into an open terminal. Are you copying in my keybinding setting when testing?

**Comment by Tyriar** [comment]
@apbendi @petecorreia do you have a custom value for `terminal.integrated.commandsToSkipShell`? You need to make sure that the commands are included.

**Comment by apbendi** [comment]
@Tyriar I'm not sure what you mean? My `keybindings.json`, in full, is:

```json
// Place your key bindings in this file to overwrite the defaults
[{
    "key": "cmd+`",
    "command": "workbench.action.terminal.focus",
    "when": "editorTextFocus"
}]
```

My `settings.json` is empty. This worked fine exactly as is before upgrading to `1.6`, with the brief exception of when it was broken by #11714

**Comment by Tyriar** [comment]
I think I may have reproduced some problems with keybindings, but I can't do it again. I'll keep this open until I can nail down a proper repro case.

Also it may be better to use `!terminalFocus`:

```js
{ "key": "cmd+`", "command": "workbench.action.terminal.focus", "when": "!terminalFocus" }
```

**Comment by Tyriar** [comment]
@apbendi I would guess @petecorreia's problem is due to a custom `terminal.integrated.commandsToSkipShell` as that's what controls keystrokes being captured by the terminal.

**Comment by Tyriar** [comment]
@alexandrudima any ideas on this? I can reproduce it (sometimes), just now my cmd+\` keybinding on Mac is just completely unfunctional, no errors in the console. I recall there's an OS-level shortcut for this, perhaps that's causing the issue?

**Comment by petecorreia** [comment]
@apbendi @Tyriar here are the bindings that are no longer working (other terminal bindings work, like cmd+t):

```js
{
    "key": "alt+cmd+up",
    "command": "workbench.action.focusFirstEditorGroup",
    "when": "terminalFocus"
}
,{
    "key": "alt+cmd+down",
    "command": "workbench.action.terminal.focus",
    "when": "editorTextFocus"
}
,{
    "key": "alt+cmd+right",
    "command": "workbench.action.terminal.focusNext",
    "when": "terminalFocus"
}
,{
    "key": "alt+cmd+left",
    "command": "workbench.action.terminal.focusPrevious",
    "when": "terminalFocus"
}
```

...and here are the commands I skip:

```js
"terminal.integrated.commandsToSkipShell": [
	"editor.action.toggleTabFocusMode",
	"workbench.action.quickOpen",
	"workbench.action.showCommands",
	"workbench.action.terminal.copySelection",
	"workbench.action.terminal.focus",
	"workbench.action.terminal.focusNext",
	"workbench.action.terminal.focusPrevious",
	"workbench.action.terminal.kill",
	"workbench.action.terminal.new",
	"workbench.action.terminal.paste",
	"workbench.action.terminal.runSelectedText",
	"workbench.action.terminal.scrollDown",
	"workbench.action.terminal.scrollUp",
	"workbench.action.terminal.toggleTerminal",
	"workbench.action.focusFirstEditorGroup"
]
```

It seems it's just the alt+cmd+(up, right, down, left) that stopped working.
Tried it on my home iMac and can reproduce.

**Comment by apbendi** [comment]
Friendly ping @Tyriar, any progress on this? Would love to have this workflow back!

**Comment by alexandrudima** [comment]
To rule out any keybinding related problem:

If you change your `keybindings.json` to:

```
// Place your key bindings in this file to overwrite the defaults
[
    { "key": "cmd+`",                "command": "actions.find" },
]
```

Does the find widget pop up / gets focus each and every time you press `"cmd+`"` ?

**Comment by Tyriar** [comment]
I think it's to do with cmd+` being an OS shortcut. People who can reproduce please follow the steps in the commend above and try using a different command.

**Comment by apbendi** [comment]
I replaced my `keybindings.json` with the exact text provided by @alexandrudima . After restarting VSCode, the new shortcut works perfectly and consistently. The find widget pops up every time. So it doesn't seem to be tied *just* to the key combo. cc  @Tyriar

**Comment by alexandrudima** [comment]
Thanks @apbendi , given your comment I believe it is not an issue with the keybindings, I'm therefore removing myself for now :).

**Comment by Tyriar** [comment]
@alexandrudima I see cmd+unknown when I use cmd+` in the define keybindings popup, that indicates an OS conflict right?

**Comment by alexandrudima** [comment]
@Tyriar I have a bug where I render `cmd` as `cmd+unknown` and I also render `cmd+[wierd unknown keycode]` as `cmd+unknown`. Created #15233 to track it. The reason is that our keybindings are represented as a number and for the case of a single `cmd`, the bit flag for the `cmd` modifier is set, meanwhile the key code is left 0, same as Unknown. You can hover over the `<input>` and you will see the last `keyCode` I've received.

I also recomment that maybe you give https://github.com/Microsoft/vscode/wiki/Troubleshooting-Keybindings a quick read such that we're all on the same page about how this works.

There you will find links to the w3c page that allows you to try out keybindings in browsers and various troubleshooting techniques.

To answer your question, it depends, `cmd+unknown` can mean we only get a keydown for `cmd` or we get a keydown for a weird key code. Hover over the `<input>` to find out.

**Comment by Tyriar** [comment]
@alexandrudima I see `keyCode: 91` both for holding <kbd>cmd</kbd> alone and holding <kbd>cmd</kbd>+<kbd>`</kbd>.

**Comment by alexandrudima** [comment]
@Tyriar Good to know. Indeed, it looks like <kbd>cmd</kbd>+<kbd>`</kbd> does not reach us whatsoever.

Using https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731(v=vs.85).aspx , `keyCode: 91 = 0x5B` maps to `VK_LWIN`, which on a mac would probably be the left <kbd>cmd</kbd> key.

**Comment by stefcameron** [comment]
I use the following similar mappings without issue in Code 1.10.1 on macOS 10.11.6 without issue:

```
  {
    "key": "ctrl+`",
    "command": "-workbench.action.terminal.toggleTerminal"
  },
  {
    "key": "cmd+`",
    "command": "workbench.action.terminal.focus",
    "when": "editorFocus" // 'focus' means show in this case
  },
  {
    "key": "cmd+`",
    "command": "workbench.action.terminal.toggleTerminal",
    "when": "terminalFocus" // 'toggle' means hide in this case
  },
```

**Comment by Tyriar** [comment]
This may have been fixed by the keycode work in 1.11 https://code.visualstudio.com/updates/v1_11#_improved-keyboard-layout-support, let me know if you're still experiencing it

**Comment by hgl** [comment]
I'm using macOS High Sierra and I'm unable bind <kbd>cmd</kbd>+<kbd>`</kbd> to anything. Pressing <kbd>cmd</kbd>+<kbd>`</kbd> in the Define Keybinding popup only displays <kbd>cmd</kbd>+

I've already disabled "Move focus to next window" option in the OS keyboard shortcuts settings, so I'm pretty sure this key combination isn't intercepted by the system.

My vscode version is `Version 1.16.0-insider (1.16.0-insider)`.

Any idea how I can fix it?

I'm able to bind this key combination in Atom though.
